packages/local_auth/android/src/main/java/io/flutter/plugins/localauth/LocalAuthPlugin.java

```diff
@@ -65,138 +85,275 @@ private LocalAuthPlugin(Activity activity) {
     }
   }
 
   /*
-   Stops the authentication if in progress.
-  */
-  private void stopAuthentication(Result result) {
+   * Returns biometric types available on device
+   */
+  private void getAvailableBiometrics(final Result result) {
     try {
-      if (authenticationHelper != null && authInProgress.get()) {
-        authenticationHelper.stopAuthentication();
-        authenticationHelper = null;
-        result.success(true);
+      Activity activity = getActivity();
+      if (activity == null || activity.isFinishing()) {
+        result.error("no_activity", "local_auth plugin requires a foreground activity", null);
         return;
       }
-      result.success(false);
+      ArrayList<String> biometrics = getAvailableBiometrics();
+      result.success(biometrics);
     } catch (Exception e) {
-      result.success(false);
+      result.error("no_biometrics_available", e.getMessage(), null);
+    }
+  }
+
+  private ArrayList<String> getAvailableBiometrics() {
+    ArrayList<String> biometrics = new ArrayList<String>();
+    Activity activity = getActivity();
+    if (activity == null || activity.isFinishing()) {
+      return biometrics;
     }
+    PackageManager packageManager = activity.getPackageManager();
+    if (Build.VERSION.SDK_INT >= 23) {
+      if (packageManager.hasSystemFeature(PackageManager.FEATURE_FINGERPRINT)) {
+        biometrics.add("fingerprint");
+      }
+    }
+    if (Build.VERSION.SDK_INT >= 29) {
+      if (packageManager.hasSystemFeature(PackageManager.FEATURE_FACE)) {
+        biometrics.add("face");
+      }
+      if (packageManager.hasSystemFeature(PackageManager.FEATURE_IRIS)) {
+        biometrics.add("iris");
+      }
+    }
+
+    return biometrics;
+  }
+
+  private boolean isDeviceSupported() {
+    KeyguardManager keyguardManager =
+        (KeyguardManager) getActivity().getBaseContext().getSystemService(KEYGUARD_SERVICE);
+    return (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M && keyguardManager.isDeviceSecure());
+  }
+
+  private boolean canAuthenticateWithBiometrics() {
+    BiometricManager biometricManager = BiometricManager.from(getActivity());
+    return biometricManager.canAuthenticate() == BiometricManager.BIOMETRIC_SUCCESS;
+  }
+
+  private boolean hasBiometricHardware() {
+    BiometricManager biometricManager = BiometricManager.from(getActivity());
+    return biometricManager.canAuthenticate() != BiometricManager.BIOMETRIC_ERROR_NO_HARDWARE;
+  }
+
+  private void isDeviceSupported(Result result) {
+    result.success(isDeviceSupported());
   }
 
   @Override
   public void onAttachedToEngine(FlutterPluginBinding binding) {
-    channel = new MethodChannel(binding.getBinaryMessenger(), CHANNEL_NAME);
+    channel = new MethodChannel(binding.getFlutterEngine().getDartExecutor(), CHANNEL_NAME);
   }
 
   @Override
   public void onDetachedFromEngine(FlutterPluginBinding binding) {}
 
+  final PluginRegistry.ActivityResultListener resultListener =
```

Comment:
Can you please move this definition up, before it is used for the first time?

Reply:
moved up